transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+			}
+		}
+		janus_mutex_unlock(&clients_mutex);
+
+		/* Start polling */
+		int res = poll(poll_fds, fds, 10000);
+		if(res == 0)
+			continue;
+		if(res < 0) {
+			JANUS_LOG(LOG_ERR, "poll() failed\n");
+			break;
+		}
+		for(int i=0; i<fds; i++) {
+			if((poll_fds[i].revents & POLLOUT) == POLLOUT) {
+				/* Find the client from its file descriptor */
+				janus_mutex_lock(&clients_mutex);
+				janus_pfunix_client *client = g_hash_table_lookup(clients_by_fd, GINT_TO_POINTER(poll_fds[i].fd));
+				if(client != NULL && client->fd > 0) {
+					char *payload = NULL;
+					while((payload = g_async_queue_try_pop(client->messages)) != NULL) {
+						int res = write(client->fd, payload, strlen(payload));
+						/* FIXME Should we check if sent everything? */
+						JANUS_LOG(LOG_HUGE, "Written %d/%zu bytes on %d\n", res, strlen(payload), client->fd);
+						g_free(payload);
+					}
+				}
+				janus_mutex_unlock(&clients_mutex);
+			}
+			if((poll_fds[i].revents & POLLIN) == POLLIN) {
+				if(poll_fds[i].fd == write_fd[0]) {
+					/* Read and ignore: we use this to unlock the poll if there's data to write */
+					res = read(poll_fds[i].fd, buffer, 4096);
+				} else if(poll_fds[i].fd == pfd) {
+					/* Janus API: accept the new client */
+					struct sockaddr_un address;
+					socklen_t address_length = 0;
+					int cfd = accept(pfd, (struct sockaddr *) &address, &address_length);
+					if(cfd > 0) {
+						JANUS_LOG(LOG_VERB, "Got new Unix Sockets Janus API client: %d\n", cfd);
+						/* Allocate new client */
+						janus_pfunix_client *client = g_malloc0(sizeof(janus_pfunix_client));
+						client->fd = cfd;
+						client->admin = FALSE;	/* Janus API client */
+						client->messages = g_async_queue_new();
+						client->session_timeout = 0;
+						/* Take note of this new client */
+						janus_mutex_lock(&clients_mutex);
+						g_hash_table_insert(clients_by_fd, GINT_TO_POINTER(cfd), client);
+						g_hash_table_insert(clients, client, client);
+						janus_mutex_unlock(&clients_mutex);
+					}
+				} else if(poll_fds[i].fd == admin_pfd) {
+					/* Admin API: accept the new client */
+					struct sockaddr_un address;
+					socklen_t address_length = 0;
+					int cfd = accept(admin_pfd, (struct sockaddr *) &address, &address_length);
+					if(cfd > 0) {
+						JANUS_LOG(LOG_VERB, "Got new Unix Sockets Admin API client: %d\n", cfd);
+						/* Allocate new client */
+						janus_pfunix_client *client = g_malloc0(sizeof(janus_pfunix_client));
+						client->fd = cfd;
+						client->admin = TRUE;	/* Admin API client */
+						client->messages = g_async_queue_new();
+						client->session_timeout = 0;
+						/* Take note of this new client */
+						janus_mutex_lock(&clients_mutex);
+						g_hash_table_insert(clients_by_fd, GINT_TO_POINTER(cfd), client);
+						g_hash_table_insert(clients, client, client);
+						janus_mutex_unlock(&clients_mutex);
+					}
+				} else {
+					/* Client data: receive */
+					res = read(poll_fds[i].fd, buffer, 4096);
```

Comment:
whatever buffer size you choose, better use sizeof, simplifies the changes if you increase it later.